File: backend/network/tcp_client.cpp
#include "tcp_client.h"
#include "../common/utils.h"
#include <iostream>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <cstring>

TCPClient::TCPClient(const std::string& host, int port)
    : host(host), port(port) {
}

void TCPClient::connect() {
    // Create the client socket
    socket = ::socket(AF_INET, SOCK_STREAM, 0);
    if (socket < 0) {
        throw std::runtime_error("Error creating socket");
    }

    // Set up the server address
    struct sockaddr_in serverAddr;
    memset(&serverAddr, 0, sizeof(serverAddr));
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(port);
    if (inet_pton(AF_INET, host.c_str(), &serverAddr.sin_addr) <= 0) {
        throw std::runtime_error("Invalid server address");
    }

    // Connect to the server
    if (::connect(socket, (struct sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        throw std::runtime_error("Error connecting to server");
    }
}

void TCPClient::send(const std::string& data) {
    Utils::writeToSocket(socket, data);
}

std::string TCPClient::receive() {
    return Utils::readFromSocket(socket);
}

File: frontend/js/tcp_client.js
import { parseResponse } from './utils.js';

class TCPClient {
    constructor(host, port) {
        this.host = host;
        this.port = port;
        this.socket = null;
    }

    async connect() {
        this.socket = new WebSocket(`ws://${this.host}:${this.port}`);
        return new Promise((resolve, reject) => {
            this.socket.onopen = () => resolve();
            this.socket.onerror = (error) => reject(error);
        });
    }

    async send(data) {
        return new Promise((resolve, reject) => {
            this.socket.onmessage = (event) => {
                const response = parseResponse(event.data);
                resolve(response);
            };
            this.socket.onerror = (error) => reject(error);
            this.socket.send(data);
        });
    }
}

export default TCPClient;